<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Data visualization with ggplot2</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">R Bootcamp for Biologists 2019</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="ion ion-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="about.html">
    <span class="ion ion-wand"></span>
     
    Instructors
  </a>
</li>
<li>
  <a href="Schedule.html">
    <span class="ion ion-calendar"></span>
     
    Schedule
  </a>
</li>
<li>
  <a href="Session0.html">
    <span class="ion ion-settings"></span>
     
    Preworkshop
  </a>
</li>
<li>
  <a href="day1.html">Day 1</a>
</li>
<li>
  <a href="day2.html">Day 2</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a>
    <span class=" favicon"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Data visualization with ggplot2</h1>

</div>


<hr />
<div class="boxobj">
<p><strong>Learning Objectives</strong></p>
<ul>
<li>Produce scatterplots, barplots, histograms, and boxplots using ggplot2.</li>
<li>Describe what faceting is and apply faceting in ggplot.</li>
<li>Modify the aesthetics of an existing ggplot plot (including axis labels and color).</li>
<li>Build customized plots from data in a data frame.</li>
<li>Be wowed by the power of R and compelled to keep using it after this workshop.</li>
</ul>
</div>
<hr />
<div id="data" class="section level2">
<h2>Data</h2>
<p>This section uses a different data set from the same experiment - the LTEE (long-term evolution experiment). This data was pulished in: Tempo and mode of genome evolution in a 50,000-generation experiment <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/">Tenaillon et al 2016</a>.</p>
<p>There are two ways to download this data file:</p>
<ol style="list-style-type: decimal">
<li>Use the <code>download.file()</code> function to download the data from the following address:</li>
</ol>
<pre class="r"><code>download.file(&quot;https://raw.githubusercontent.com/alanaschick/Rbootcamp2019/master/data/combined_tidy_vcf.csv&quot;, &quot;data/combined_tidy_vcf.csv&quot;)</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Download the data onto your computer from <a href="">this</a> dropbox link and move it into the <code>data</code> directory of your RStudio project.</li>
</ol>
<div class="box">
<h3 id="exercise">EXERCISE</h3>
<p>Read the data into R. Call this data <code>variants</code>. Ensure that you have read the data in by calling <code>head(variants)</code>.</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- variants <- read.csv("data/combined_tidy_vcf.csv") -->
<!-- ``` -->
<p>You can further investigate the structure of the data frame using the <code>str()</code> function:</p>
<pre class="r"><code>str(variants)</code></pre>
<pre><code>## &#39;data.frame&#39;:    801 obs. of  29 variables:
##  $ sample_id    : Factor w/ 3 levels &quot;SRR2584863&quot;,&quot;SRR2584866&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ CHROM        : Factor w/ 1 level &quot;CP000819.1&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##  $ POS          : int  9972 263235 281923 433359 473901 648692 1331794 1733343 2103887 2333538 ...
##  $ ID           : logi  NA NA NA NA NA NA ...
##  $ REF          : Factor w/ 59 levels &quot;A&quot;,&quot;ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG&quot;,..: 49 33 33 30 24 16 16 33 2 12 ...
##  $ ALT          : Factor w/ 57 levels &quot;A&quot;,&quot;AC&quot;,&quot;ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG&quot;,..: 31 46 46 29 25 46 1 1 4 15 ...
##  $ QUAL         : num  91 85 217 64 228 210 178 225 56 167 ...
##  $ FILTER       : logi  NA NA NA NA NA NA ...
##  $ INDEL        : logi  FALSE FALSE FALSE TRUE TRUE FALSE ...
##  $ IDV          : int  NA NA NA 12 9 NA NA NA 2 7 ...
##  $ IMF          : num  NA NA NA 1 0.9 ...
##  $ DP           : int  4 6 10 12 10 10 8 11 3 7 ...
##  $ VDB          : num  0.0257 0.0961 0.7741 0.4777 0.6595 ...
##  $ RPB          : num  NA 1 NA NA NA NA NA NA NA NA ...
##  $ MQB          : num  NA 1 NA NA NA NA NA NA NA NA ...
##  $ BQB          : num  NA 1 NA NA NA NA NA NA NA NA ...
##  $ MQSB         : num  NA NA 0.975 1 0.916 ...
##  $ SGB          : num  -0.556 -0.591 -0.662 -0.676 -0.662 ...
##  $ MQ0F         : num  0 0.167 0 0 0 ...
##  $ ICB          : logi  NA NA NA NA NA NA ...
##  $ HOB          : logi  NA NA NA NA NA NA ...
##  $ AC           : int  1 1 1 1 1 1 1 1 1 1 ...
##  $ AN           : int  1 1 1 1 1 1 1 1 1 1 ...
##  $ DP4          : Factor w/ 217 levels &quot;0,0,0,2&quot;,&quot;0,0,0,3&quot;,..: 3 132 73 141 176 104 61 74 133 137 ...
##  $ MQ           : int  60 33 60 60 60 60 60 60 60 60 ...
##  $ Indiv        : Factor w/ 3 levels &quot;/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ gt_PL        : Factor w/ 206 levels &quot;100,0&quot;,&quot;103,0&quot;,..: 16 10 134 198 142 127 93 142 9 80 ...
##  $ gt_GT        : int  1 1 1 1 1 1 1 1 1 1 ...
##  $ gt_GT_alleles: Factor w/ 57 levels &quot;A&quot;,&quot;AC&quot;,&quot;ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG&quot;,..: 31 46 46 29 25 46 1 1 4 15 ...</code></pre>
<div class="box">
<h3 id="exercise-1">EXERCISE</h3>
<p>Take a few minutes to familiarize yourself with this dataset. There are a lot of variables (how many?), so only worry about the ones listed below.</p>
<table style="width:71%;">
<colgroup>
<col width="6%" />
<col width="63%" />
</colgroup>
<thead>
<tr class="header">
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>sample_id</td>
<td>sample ID</td>
</tr>
<tr class="even">
<td>CHROM</td>
<td>contig location where the variation occurs</td>
</tr>
<tr class="odd">
<td>POS</td>
<td>position within the contig where the variation occurs</td>
</tr>
<tr class="even">
<td>REF</td>
<td>reference genotype (forward strand)</td>
</tr>
<tr class="odd">
<td>ALT</td>
<td>sample genotype (forward strand)</td>
</tr>
<tr class="even">
<td>QUAL</td>
<td>Phred-scaled probablity that the observed variant exists at this site (higher is better)</td>
</tr>
<tr class="odd">
<td>INDEL</td>
<td>whether the variant is an indel</td>
</tr>
<tr class="even">
<td>IDV</td>
<td>length of indel</td>
</tr>
<tr class="odd">
<td>IMF</td>
<td>maximum fraction of reads supporting an indel</td>
</tr>
<tr class="even">
<td>DP</td>
<td>the depth per allele by sample and coverage</td>
</tr>
<tr class="odd">
<td>MQ</td>
<td>mapping quality</td>
</tr>
<tr class="even">
<td>Indiv</td>
<td>name of file</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="plotting-with-ggplot2" class="section level1">
<h1>Plotting with ggplot2</h1>
<p>We start by loading the <code>ggplot2</code> package:</p>
<pre class="r"><code>library(ggplot2)</code></pre>
<p><code>ggplot2</code> is a plotting package that makes it simple to create complex plots from data in a data frame. It provides a more programmatic interface for specifying what variables to plot, how they are displayed, and general visual properties. Therefore, we only need minimal changes if the underlying data change or if we decide to change from a bar plot to a scatter plot. This helps in creating publication quality plots with minimal amounts of adjustments and tweaking.</p>
<p><code>ggplot2</code> functions like data in the ‘long’ format, i.e., a column for every dimension, and a row for every observation. Well-structured data will save you lots of time when making figures with <code>ggplot2</code>.</p>
<p>Graphics are built step by step by adding new elements. Adding layers in this fashion allows for extensive flexibility and customization of plots.</p>
<p>To build a ggplot, we will use the following basic template that can be used for different types of plots:</p>
<pre class="r"><code>ggplot(data = &lt;DATA&gt;, mapping = aes(&lt;MAPPINGS&gt;)) +  &lt;GEOM_FUNCTION&gt;()</code></pre>
<ol style="list-style-type: decimal">
<li>Specify which data set to use for the plot using the <code>data</code> argument:</li>
</ol>
<pre class="r"><code>ggplot(data = variants)</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Define a “mapping” (using the aesthetic (<code>aes</code>) function), by selecting the variables to be plotted and specifying how to present them in the graph, e.g. as x/y positions or characteristics such as size, shape, color, etc:</li>
</ol>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = DP))</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Add “geoms” – graphical representations of the data in the plot (points, lines, bars). <code>ggplot2</code> offers many different geoms; we will use some common ones today, including:</li>
</ol>
<ul>
<li><code>geom_point()</code> for scatter plots, dot plots, etc.</li>
<li><code>geom_boxplot()</code> for boxplots.</li>
<li><code>geom_histogram()</code> for histograms.</li>
<li><code>geom_barplot()</code> for barplots.</li>
<li><code>geom_line()</code> for trend lines, time series, etc.</li>
</ul>
<p>To add a geom to the plot use the <code>+</code> operator. Because we have two continuous variables, let’s use <code>geom_point()</code> first:</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = DP)) + 
  geom_point()</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>The <code>+</code> in the <code>ggplot2</code> package is particularly useful because it allows you to modify existing <code>ggplot</code> objects. This means you can easily set up plot templates and conveniently explore different types of plots, so the above plot can also be generated with code like this:</p>
<pre class="r"><code># Assign plot to a variable
coverage_plot &lt;- ggplot(data = variants, aes(x = POS, y = DP))

# Draw the plot
coverage_plot + 
    geom_point()</code></pre>
<p>Notes:</p>
<ul>
<li>Anything you put in the <code>ggplot()</code> function can be seen by any geom layers that you add (i.e., these are universal plot settings). This includes the x- and y-axis mapping you set up in <code>aes()</code>.</li>
<li>You can also specify mappings for a given geom independently of the mappings defined globally in the <code>ggplot()</code> function.</li>
<li>The <code>+</code> sign used to add new layers must be placed at the end of the line containing the previous layer. If, instead, the <code>+</code> sign is added at the beginning of the line containing the new layer, <code>ggplot2</code> will not add the new layer and will return an error message.</li>
</ul>
<pre class="r"><code># This is the correct syntax for adding layers
coverage_plot +
  geom_point()

# This will not add the new layer and will return an error message
coverage_plot
  + geom_point()</code></pre>
<div class="boxGrey">
<h2 id="principles-of-effective-display">Principles of effective display</h2>
<p>SOURCE: (Whitlock &amp; Schluter, The Analysis of Biological Data)[<a href="http://whitlockschluter.zoology.ubc.ca/" class="uri">http://whitlockschluter.zoology.ubc.ca/</a>]</p>
<p>We will follow these metrics to create and evaluate figures:<br />
1. Show the data<br />
2. Make patterns in the data easy to see<br />
3. Represent magnitudes honestly<br />
4. Draw graphical elements clearly, minimizing clutter</p>
</div>
<div class="box">
<h3 id="exercise-2">EXERCISE</h3>
<p>Create a scatter plot (using the <code>geom_point()</code> function for quality (<code>QUAL</code>) versus coverage depth (<code>DP</code>).</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = QUAL, y = DP)) +  -->
<!--   geom_point() -->
<!-- ``` -->
</div>
<div id="building-plots-iteratively" class="section level1">
<h1>Building plots iteratively</h1>
<p>Building plots with <code>ggplot2</code> is typically an iterative process. We start by defining the dataset we’ll use, lay out the axes, and choose a geom:</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = DP)) + 
  geom_point()</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>Then, we start modifying this plot to extract more information from it. For instance, we can add transparency (<code>alpha</code>) to avoid overplotting:</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = DP)) +
    geom_point(alpha = 0.5)</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>We can also add colors for all the points:</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = DP)) +
  geom_point(alpha = 0.5, color = &quot;blue&quot;)</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>Or to color each species in the plot differently, you could use a vector as an input to the argument color. <code>ggplot2</code> will provide a different color corresponding to different values in the vector. Here is an example where we color with <code>sample_id</code>:</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = DP, color = sample_id)) + 
  geom_point(alpha = 0.5)</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>Notice that we can change the geom layer and colors will be still determined by <code>sample_id</code>:</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = DP, color = sample_id)) + 
  geom_jitter(alpha = 0.5)</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>To make our plot more readable, we can add axis labels:</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = DP, color = sample_id)) + 
  geom_jitter(alpha = 0.5) + 
  labs(x = &quot;Base Pair Position&quot;, 
       y = &quot;Read Depth (DP)&quot;)</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<div class="box">
<h3 id="exercise-3">EXERCISE</h3>
<p>Create a scatter plot of mapping quality (<code>MQ</code>) over position (<code>POS</code>) with the samples showing in different colors. Make sure to give your plot relevant axis labels.</p>
</div>
<!-- Solution -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = POS, y = MQ, color = sample_id)) + -->
<!--   geom_point() + -->
<!--   labs(x = "Base Pair Position", -->
<!--        y = "Mapping Quality (MQ)") -->
<!-- ``` -->
</div>
<div id="faceting" class="section level1">
<h1>Faceting</h1>
<p><code>ggplot2</code> has a special technique called faceting that allows the user to split one plot into multiple plots based on a factor included in the dataset. We will use it to split our mapping quality plot into three panels, one for each sample.</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = MQ, color = sample_id)) + 
 geom_point() + 
 labs(x = &quot;Base Pair Position&quot;, 
      y = &quot;Mapping Quality (MQ)&quot;) + 
 facet_grid(. ~ sample_id)</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>This looks ok, but it would be easier to read if the plot facets were stacked vertically rather than horizontally. The <code>facet_grid</code> geometry allows you to explicitly specify how you want your plots to be arranged via formula notation (<code>rows ~ columns</code>); a <code>.</code> can be used as a placeholder that indicates only one row or column).</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = MQ, color = sample_id)) + 
 geom_point() + 
 labs(x = &quot;Base Pair Position&quot;, 
      y = &quot;Mapping Quality (MQ)&quot;) + 
 facet_grid(sample_id ~ .)</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
<p>Usually plots with white background look more readable when printed. We can set the background to white using the function <code>theme_bw()</code>. Additionally, you can remove the grid:</p>
<pre class="r"><code>ggplot(data = variants, aes(x = POS, y = MQ, color = sample_id)) + 
  geom_point() + 
  labs(x = &quot;Base Pair Position&quot;, 
       y = &quot;Mapping Quality (MQ)&quot;) + 
  facet_grid(sample_id ~ .) +
  theme_bw() +
  theme(panel.grid = element_blank())</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<div class="box">
<h3 id="exercise-4">EXERCISE</h3>
<p>Use what you just learned to create a scatter plot of PHRED scaled quality (<code>QUAL</code>) over position (<code>POS</code>) with the samples showing in different colors. Make sure to give your plot relevant axis labels.</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = POS, y = QUAL, color = sample_id)) + -->
<!--   geom_point() + -->
<!--   labs(x = "Base Pair Position", -->
<!--        y = "PHRED-sacled Quality (QUAL)") + -->
<!--   facet_grid(sample_id ~ .) -->
<!-- ``` -->
</div>
<div id="barplots" class="section level1">
<h1>Barplots</h1>
<p>We can create barplots using the <code>geom_bar</code> geom. Let’s make a barplot showing the number of variants for each sample that are indels.</p>
<pre class="r"><code>ggplot(data = variants, aes(x = INDEL, fill = sample_id)) + 
  geom_bar() +
  facet_grid(sample_id ~ .)</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
<div class="box">
<h3 id="exercise-5">EXERCISE</h3>
<p>Since we already have the sample_id labels on the individual plot facets, we don’t need the legend. Use the help file for <code>geom_bar</code> and any other online resources you want to use to remove the legend from the plot.</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = INDEL, color = sample_id)) +  -->
<!--    geom_bar(show.legend = F) + -->
<!--    facet_grid(sample_id ~ .) -->
<!-- ``` -->
</div>
<div id="histograms" class="section level1">
<h1>Histograms</h1>
<p>Sometimes it can be useful to plot a single variable at a time. Usually this is for exploratory purposes - to get a feel for a variable. To do this, use the <code>geom_histogram()</code> function. For example, to look at the distribution of Qualities:</p>
<pre class="r"><code>ggplot(data = variants, aes(QUAL)) +
  geom_histogram()</code></pre>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<p>R is giving us a warning. We can choose to ignore this if we are able to see what we want in the plot. Otherwise, we can change the <code>binwidth</code> by using the binwidth argument:</p>
<pre class="r"><code>ggplot(data = variants, aes(QUAL)) +
  geom_histogram(binwidth = 5)</code></pre>
<img src="07-ggplot2_files/figure-html/unnamed-chunk-21-1.png" width="672" />
<div class="box">
<h3 id="exercise-6">EXERCISE</h3>
<p>Create a plot that shows the distribution of Read Depth (DP) for each sample separately.</p>
<p><b>Bonus challenge.</b> Create this plot for only two of the three samples IDs - the two with many fewer variants.</p>
</div>
<!-- ### Solution -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(DP)) + -->
<!--   geom_histogram(binwidth = 2) + -->
<!--   facet_grid(sample_id ~ .) -->
<!-- ``` -->
<!-- ### Bonus solution -->
<!-- ```{r} -->
<!-- # subset the data -->
<!-- two <- filter(variants, sample_id != "SRR2584866") -->
<!-- ggplot(data = two, aes(DP)) + -->
<!--   geom_histogram(binwidth = 2) + -->
<!--   facet_grid(sample_id ~ .) -->
<!-- ``` -->
</div>
<div id="boxplots" class="section level1">
<h1>Boxplots</h1>
<p>When there are a large number of values for a certain variable (like for one of the samples above), boxplots can be a useful way to display summary statistics like the median, and “spread” of a variable.</p>
<div id="the-five-number-summary" class="section level3">
<h3>The Five Number Summary</h3>
<p>The five number summary gives a quick look at the features of numerical variables. It consists of the variables:</p>
<ul>
<li>minimum</li>
<li>1st quartile</li>
<li>median</li>
<li>3rd quartile</li>
<li>maximum</li>
</ul>
<div class="boxSmall">
<p><strong>QUANTILES:</strong> The <em>pth</em> percentile of a data set sorted from smallest to largest is the value such that <em>p</em> percent of the data are at or below this value. The quartiles are special percentiles; the 1st quartile is the 25th percentile, and the 3rd quartile is the 75th percentile. The median is also a quartile – it is the 50th percentile.</p>
</div>
<p>Within these five numbers is a lot of useful data!</p>
<ul>
<li>the median gives a measure of the center of the data</li>
<li>the minimum and maximum give the range of the data</li>
<li>the 1st and 3rd quartiles give a sense of the spread of the data, especially when compared to the minimum, maximum, and median</li>
</ul>
<p>To create a boxplot, use the <code>geom_boxplot()</code> function:</p>
<pre class="r"><code>ggplot(data = variants, aes(x = sample_id, y = DP)) + 
  geom_boxplot()</code></pre>
<p><img src="07-ggplot2_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<div class="box">
<h3 id="exercise-7">EXERCISE</h3>
<ol style="list-style-type: decimal">
<li>Change the colour of each box in the above plot to match the sample_id colours in the plots above.</li>
<li>Log transform the y-axis.</li>
</ol>
<p><b>Bonus challenge:</b> Change the name of the legend to “Sample ID”.</p>
</div>
<!-- ### SOLUTION -->
<!-- ```{r} -->
<!-- ggplot(data = variants, aes(x = sample_id, y = log(DP), color = sample_id)) + -->
<!--   geom_boxplot() -->
<!-- ``` -->
</div>
</div>
<div id="themes" class="section level1">
<h1>Themes</h1>
<p>In addition to <code>theme_bw()</code>, which changes the plot background to white, <code>ggplot2</code> comes with several other themes which can be useful to quickly change the look of your visualization. The complete list of themes is available at <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="uri">https://ggplot2.tidyverse.org/reference/ggtheme.html</a>. <code>theme_minimal()</code> and <code>theme_light()</code> are popular, and <code>theme_void()</code> can be useful as a starting point to create a new hand-crafted theme.</p>
<p>The <code>ggthemes</code> package provides a wide variety of options (including an Excel 2003 theme). The <code>ggplot2</code> <a href="https://www.ggplot2-exts.org">extensions website</a> provides a list of packages that extend the capabilities of <code>ggplot2</code>, including additional themes.</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
